--- morse_overrides.sh	2026-05-20 10:26:01.000000000 +0900
+++ morse_overrides.sh.new	2026-09-22 04:53:44.571612703 +0900
@@ -751,6 +751,7 @@
 	morse_override_wpa3_encryption
 
 	json_get_vars \
+		iw_enabled iw_rcois iw_realm iw_domain_name \
 		ssid bssid key \
 		basic_rate mcast_rate \
 		ieee80211w ieee80211r fils ocv \
@@ -782,6 +783,7 @@
 
 	local key_mgmt='NONE'
 	local network_data=
+	local cred_data=
 	local T="	"
 
 	local scan_ssid="scan_ssid=1"
@@ -899,13 +901,20 @@
 
 			json_get_vars eap_type identity anonymous_identity ca_cert ca_cert_usesystem
 
+			[ -n "$iw_rcois" ] && append cred_data "roaming_consortiums=\"$iw_rcois\"" "$N$T"
+			[ -n "$iw_realm" ] && append cred_data "realm=\"$iw_realm\"" "$N$T"
+			json_for_each_item append_iw_domain iw_domain_name
+
 			[ "$fils" -gt 0 ] && append network_data "erp=1" "$N$T"
 			if [ "$ca_cert_usesystem" -eq "1" -a -f "/etc/ssl/certs/ca-certificates.crt" ]; then
 				append network_data "ca_cert=\"/etc/ssl/certs/ca-certificates.crt\"" "$N$T"
+				append cred_data "ca_cert=\"/etc/ssl/certs/ca-certificates.crt\"" "$N$T"
 			else
 				[ -n "$ca_cert" ] && append network_data "ca_cert=\"$ca_cert\"" "$N$T"
+				[ -n "$ca_cert" ] && append cred_data "ca_cert=\"$ca_cert\"" "$N$T"
 			fi
 			[ -n "$identity" ] && append network_data "identity=\"$identity\"" "$N$T"
+			[ -n "$identity" ] && append cred_data "username=\"$identity\"" "$N$T"
 			[ -n "$anonymous_identity" ] && append network_data "anonymous_identity=\"$anonymous_identity\"" "$N$T"
 			case "$eap_type" in
 				tls)
@@ -924,6 +933,9 @@
 					fi
 					append network_data "private_key=\"$priv_key\"" "$N$T"
 					append network_data "private_key_passwd=\"$priv_key_pwd\"" "$N$T"
+					[ -n "$client_cert" ] && append cred_data "client_cert=\"$client_cert\"" "$N$T"
+					append cred_data "private_key=\"$priv_key\"" "$N$T"
+					append cred_data "private_key_passwd=\"$priv_key_pwd\"" "$N$T"
 
 					json_get_vars subject_match
 					[ -n "$subject_match" ] && append network_data "subject_match=\"$subject_match\"" "$N$T"
@@ -953,6 +965,7 @@
 							append list "$x" ";"
 						done
 						append network_data "domain_suffix_match=\"$list\"" "$N$T"
+						append cred_data "domain_suffix_match=\"$list\"" "$N$T"
 					fi
 				;;
 				fast|peap|ttls)
@@ -970,6 +983,7 @@
 						append network_data "private_key2_passwd=\"$priv_key2_pwd\"" "$N$T"
 					else
 						append network_data "password=\"$password\"" "$N$T"
+						append cred_data "password=\"$password\"" "$N$T"
 					fi
 
 					json_get_vars subject_match
@@ -991,6 +1005,7 @@
 							append list "$x" ";"
 						done
 						append network_data "domain_match=\"$list\"" "$N$T"
+						append cred_data "domain_match=\"$list\"" "$N$T"
 					fi
 
 					json_get_values domain_suffix_match domain_suffix_match
@@ -1000,6 +1015,7 @@
 							append list "$x" ";"
 						done
 						append network_data "domain_suffix_match=\"$list\"" "$N$T"
+						append cred_data "domain_suffix_match=\"$list\"" "$N$T"
 					fi
 
 					phase2proto="auth="
@@ -1043,10 +1059,12 @@
 						;;
 					esac
 					append network_data "phase2=\"$phase2proto$auth\"" "$N$T"
+					append cred_data "phase2=\"$phase2proto$auth\"" "$N$T"
 				;;
 			esac
 			eap_type_data=$(echo $eap_type | tr 'a-z' 'A-Z')
 			append network_data "eap=$eap_type_data" "$N$T"
+			append cred_data "eap=$eap_type_data" "$N$T"
 		;;
 	esac
 
@@ -1071,6 +1089,7 @@
 	case "$ieee80211w" in
 		[012])
 			[ "$wpa" -ge 2 ] && append network_data "ieee80211w=$ieee80211w" "$N$T"
+			pmf="pmf=$ieee80211w"
 		;;
 	esac
 
@@ -1123,7 +1142,7 @@
 		if [ -z "$ssid" ]; then
 			echo "No creds in UCI to update network config"
 		else
-			wpa_add_network_block "multiap_bh" "$_config"
+			wpa_add_network_block "multiap_bh" "$_config" "$iw_enabled"
 		fi
 	elif [ "$matter_enable" = 1 ]; then
 		echo "update_config=1"  >> "$_config"
@@ -1162,7 +1181,7 @@
 			append network_data "mac_value=$(cat "/sys/class/net/$network_ifname/address")" "$N$T"
 		fi
 
-		wpa_add_network_block "sta" "$_config"
+		wpa_add_network_block "sta" "$_config" "$iw_enabled"
 	fi
 	return 0
 }
@@ -1170,13 +1189,27 @@
 
 # $1: network type (sta, multiap_bh, ...)
 # $2: output file
+# $3: Interworking enabled or not
 wpa_add_network_block()
 {
 	local network_type=$1
 	local output_file=$2
+	local iw_enabled=$3
 	case "$network_type" in
 		sta)
-			cat >> "$output_file" <<EOF
+			if [ "$iw_enabled" = "1" ]; then
+				cat >> "$output_file" <<EOF
+interworking=1
+hs20=1
+auto_interworking=1
+$pmf
+
+cred={
+	$cred_data
+}
+EOF
+			else
+				cat >> "$output_file" <<EOF
 $mesh_data
 network={
 	$scan_ssid
@@ -1186,6 +1219,7 @@
 	$twt_block
 }
 EOF
+			fi
 		;;
 		multiap_bh)
 			cat >> "$_config" <<EOF
